prometheus/exporter/job_exporter.py

```diff
@@ -96,22 +101,55 @@ def gen_job_metrics(logDir, gpuMetrics):
         outputFile.write(containerBlockOut)
         outputFile.write(containerMemPerc)
 
+def try_get_gpu_metrics(semaphore, logDir):
+    """ gen_gpu_metrics_from_smi may block indefinitely, so we wrap call in thread.
+    Also, to avoid having too much threads, use semaphore to ensure only 1 thread is running """
+    if semaphore.acquire(False):
+        def wrapper(queue):
+            """ wrapper assume semaphore already acquired, will release semaphore on exit """
+            result = None
+            start = datetime.datetime.now()
+
+            try:
+                result = gpu_exporter.gen_gpu_metrics_from_smi(logDir)
+            except Exception as e:
+                logger.warn("get gpu metrics failed")
+                logger.exception(e)
+            finally:
+                logger.info("get gpu metrics spent %s", datetime.datetime.now() - start)
+                semaphore.release()
+                queue.put(result)
+
+        queue = Queue(1)
+        t = threading.Thread(target=wrapper, args=(queue,), name="gpu-metrices-getter")
+        t.start()
+        try:
+            result = queue.get(block=True, timeout=3)
+            return result
+        except Empty:
+            return None
+    else:
+        logger.warn("another thread is running")
+
+    return None
+
 def main(argv):
     logDir = argv[0]
     timeSleep = int(argv[1])
     iter = 0
-    while(True):
+
+    semaphore = threading.Semaphore(1)
+
+    while True:
         try:
             logger.info("job exporter running {0} iteration".format(str(iter)))
             iter += 1
             # collect GPU metrics
-            gpuMetrics = gpu_exporter.gen_gpu_metrics_from_smi(logDir)
+            gpuMetrics = try_get_gpu_metrics(semaphore, logDir)
```

Comment:
1 consideration: Suggest judge the nvidia-smi launch time and drop  nvidia-smi long duration ago old metrics? if it is hanging long time then it return results after duration d [Could consider or set a duration threshold ]. The metric is the older gpu metrics, maybe mistake operator?